### PR TITLE
feat(groq): Deterministic emoji weather forecast utility that maps any date to a whimsical weather emoji without external APIs.

### PR DESCRIPTION
**Pull request by polsala**
## Implementation Summary
- **Utility**: emoji-forecast
- **Provider**: groq
- **Location**: `utils/nightly-emoji-forecast`
- **Files Created**: 3
- **Description**: Deterministic emoji weather forecast utility that maps any date to a whimsical weather emoji without external APIs.

## Rationale
- Automated proposal from the Groq generator delivering a fresh community utility.
- This utility was generated using the **groq** AI provider.

## Why safe to merge
- Utility is isolated to `utils/nightly-emoji-forecast`.
- README + tests ship together (see folder contents).
- No secrets or credentials touched.
- All changes are additive and self-contained.

## Test Plan
- Follow the instructions in the generated README at `utils/nightly-emoji-forecast/README.md`
- Run tests located in `utils/nightly-emoji-forecast/tests/`

## Links
- Generated docs and examples committed alongside this change.

## Mock Justification
- Not applicable; generator did not introduce new mocks.

### CODE REVIEW

**Comment by polsala** [comment]
## 🤖 Review by GEMINI Agent

✅ What's solid
*   **Clear Purpose and Implementation:** The utility's goal of providing a deterministic emoji weather forecast is well-defined, and the `get_forecast` function implements this logic clearly and concisely.
*   **Self-Contained and Additive:** The utility is isolated, additive, and self-contained, minimizing potential side effects on the existing codebase.
*   **Robust CLI:** The `main` function provides a user-friendly command-line interface with clear usage instructions and error handling for missing arguments and invalid date formats.
*   **Comprehensive Documentation:** The `README.md` clearly explains the utility's purpose, usage, and the deterministic algorithm behind it, which is excellent for developer experience.

🧪 Tests
*   **Deterministic Coverage:** The tests effectively cover the deterministic nature of the `get_forecast` function by asserting expected emoji outputs for specific dates, including edge cases within the modulo cycle.
*   **Relative Import:** The use of `from ..src.forecast import get_forecast` for importing the function from a sibling package is correct and demonstrates good Python package structure awareness for testing.
*   **Test Naming:** Test method names like `test_known_dates` and `test_mocked_today` are descriptive and clearly indicate their purpose.

**Actionable Feedback for Tests:**
*   **Clarify "Mock Rationale":** The comment in `test_mocked_today` states "Mock rationale: directly passing a date object avoids any external calls." While passing a `date` object is good practice for testing pure functions, it's not strictly "mocking" in the traditional sense (e.g., using `unittest.mock.patch`). Consider rephrasing this comment to clarify that the function's design inherently avoids external dependencies, making direct testing straightforward without the need for complex mocking.
*   **Expand Leap Year Coverage:** While the `day_of_year` logic implicitly handles leap years (Feb 29 would be day 60), adding a specific test case for a date in a leap year (e.g., `date(2024, 2, 29)`) could explicitly confirm this behavior.

🔒 Security
*   **No External Dependencies/APIs:** The utility's core logic is entirely self-contained and deterministic, relying solely on date calculations. This eliminates risks associated with external API calls, network vulnerabilities, or third-party data breaches.
*   **No Secrets/Credentials:** As stated in the PR body, no secrets or credentials are touched, which is a significant security positive.
*   **Input Validation:** The `main` function includes basic input validation for the date format, preventing common parsing errors.

**Actionable Feedback for Security:**
*   **Consider `sys.exit` in `main` for Library Context:** While `sys.exit(1)` is standard for CLI errors, if this utility were ever intended to be imported and used programmatically as part of a larger application, raising specific exceptions (e.g., `ValueError`) instead of calling `sys.exit` would allow calling code to handle errors gracefully without terminating the entire process. For a standalone utility, the current approach is acceptable.

🧩 Docs/DX
*   **Clear README:** The `README.md` is well-structured, easy to understand, and provides all necessary information for a developer to quickly grasp and use the utility. The "How it works" section is particularly helpful.
*   **Type Hinting:** The `get_forecast` function uses type hints (`target_date: date) -> str`), which significantly improves code readability and maintainability for developers.
*   **Docstrings:** Both the module and the `get_forecast` function have clear and concise docstrings, aiding in code comprehension.

**Actionable Feedback for Docs/DX:**
*   **Deeply Nested Directory Structure:** The most significant DX issue is the deeply nested and redundant directory structure: `utils/nightly-emoji-forecast/utils/nightly-emoji-forecast/utils/emoji-forecast/`. This makes the utility harder to locate, import, and manage. It should ideally be flattened to a more conventional structure, such as `utils/emoji-forecast/`.

    ```
    # Current structure:
    # utils/nightly-emoji-forecast/
    # └── utils/
    #     └── nightly-emoji-forecast/
    #         └── utils/
    #             └── emoji-forecast/
    #                 ├── README.md
    #                 ├── src/
    #                 │   └── forecast.py
    #                 └── tests/
    #                     └── test_forecast.py

    # Suggested simplified structure:
    # utils/emoji-forecast/
    # ├── README.md
    # ├── src/
    # │   └── forecast.py
    # └── tests/
    #     └── test_forecast.py
    ```
*   **Consistency in Cycle Description:** In the `README.md`, the "How it works" section accurately describes the `day_of_year % 4` logic. However, the initial listing of the `WEATHER_CYCLE` as `☀️` (sun), `⛅` (partly cloudy), `🌧️` (rain), `❄️` (snow) might lead a reader to assume `☀️` is for day 1, which contradicts the actual implementation (day 1 is `⛅`). Consider rephrasing the cycle description to explicitly state the mapping or adjust the `WEATHER_CYCLE` definition if `☀️` is intended for day 1. For example, `WEATHER_CYCLE = ["☀️", "⛅", "🌧️", "❄️"]` maps `day_of_year % 4 = 0` to `☀️`, `1` to `⛅`, `2` to `🌧️`, `3` to `❄️`.

    ```diff
    --- a/utils/nightly-emoji-forecast/utils/nightly-emoji-forecast/utils/emoji-forecast/README.md
    +++ b/utils/nightly-emoji-forecast/utils/nightly-emoji-forecast/utils/emoji-forecast/README.md
    @@ -12,7 +12,7 @@
     ```
     
     ## How it works
    -- The utility defines a four‑item cycle: `☀️` (sun), `⛅` (partly cloudy), `🌧️` (rain), `❄️` (snow).
    +- The utility defines a four‑item cycle, mapped by `day_of_year % 4`: `☀️` (for 0), `⛅` (for 1), `🌧️` (for 2), `❄️` (for 3).
     - The day of the year (`1`‑`366`) is taken modulo `4` to pick an emoji from the cycle.
     - Because the algorithm is pure math, the same date always yields the same emoji.
    ```

🧱 Mocks/Fakes
*   **No External Dependencies:** The `get_forecast` function is a pure function, taking a `date` object as input and returning a string. This design inherently avoids the need for complex mocking or faking of external systems (like APIs or databases).
*   **Direct Testing:** The tests directly pass `date` objects, which is the most straightforward and reliable way to test such a pure function.

**Actionable Feedback for Mocks/Fakes:**
*   **Clarify "Mock Justification" in PR Body:** The PR body states "Mock Justification: Not applicable; generator did not introduce new mocks." While technically true for *new* mocks, the `test_mocked_today` comment in the test file itself uses the term "mock rationale." To avoid confusion, consider rephrasing the PR body's statement to align with the actual testing approach, for example: "Not applicable; the utility's pure function design allows for direct testing without external dependencies or complex mocking." This reinforces the positive design choice.

**Comment by polsala** [comment]
## 🤖 Review by OPENROUTER Agent

✅ What's solid
- Clear, self-contained utility with a well-defined purpose: deterministic emoji weather mapping.
- Implementation is simple and correct: uses day-of-year modulo 4 to select from a fixed emoji cycle.
- README provides concise usage instructions and explains the deterministic logic.
- Tests are deterministic and cover known dates and edge cases (e.g., Dec 31).

🧪 Tests
- Tests are appropriately deterministic and verify the core logic.
- Consider adding a test for a leap year date (e.g., Feb 29) to ensure day-of-year calculation works correctly in leap years.
- Add a test for a date that wraps around the year (e.g., Dec 31) to confirm modulo behavior across year boundaries.

🔒 Security
- No security concerns identified; the utility is purely computational and does not interact with external systems or user input beyond date parsing.
- Date parsing uses `strptime` with a strict format, which is safe from injection risks.

🧩 Docs/DX
- README is clear but could benefit from a brief explanation of the emoji meanings (sun, partly cloudy, rain, snow) for users unfamiliar with the symbols.
- Consider adding a note about leap year handling to clarify that Feb 29 is treated as day 60 and included in the cycle.

🧱 Mocks/Fakes
- No mocks are needed for this utility as it is purely computational and does not depend on external services or state.

**Review by polsalaTipoDeIncognito — APPROVED** [review]
✅ Auto-approved: All AI agent reviews passed and CI checks succeeded.
